src/main/java/net/minecraftforge/registries/ForgeRegistry.java

```diff
@@ -871,6 +883,45 @@ public Snapshot makeSnapshot()
         return ret;
     }
 
+    private class RegistryCodec implements Codec<V>
+    {
+        @Override
+        public <T> DataResult<Pair<V, T>> decode(DynamicOps<T> ops, T input)
+        {
+            if(ops.compressMaps())
+            {
+                return ops.getNumberValue(input).flatMap(n ->
+                {
+                    int id = n.intValue();
+                    if(ids.get(id) == null)
+                    {
+                        return DataResult.error("Unknown registry id " + n);
+                    }
+                    V val = ForgeRegistry.this.getValue(id);
+                    return DataResult.success(val);
+                }).map(v -> Pair.of(v, ops.empty()));
+            }
+            else
+            {
+                return ResourceLocation.CODEC.decode(ops, input).flatMap(keyValuePair -> !ForgeRegistry.this.containsKey(keyValuePair.getFirst()) ?
+                        DataResult.error("Unknown registry key: " + keyValuePair.getFirst()) :
+                        DataResult.success(keyValuePair.mapFirst(ForgeRegistry.this::getValue)));
```

Comment:
I suggest moving the ternary operators a bit to make it more readable:
```suggestion
                return ResourceLocation.CODEC.decode(ops, input).flatMap(keyValuePair -> !ForgeRegistry.this.containsKey(keyValuePair.getFirst())
                        ? DataResult.error("Unknown registry key: " + keyValuePair.getFirst())
                        : DataResult.success(keyValuePair.mapFirst(ForgeRegistry.this::getValue)));
```